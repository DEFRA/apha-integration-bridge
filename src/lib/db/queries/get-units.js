import Joi from 'joi'

import { query } from '../operations/query.js'

export const GetUnitsSchema = Joi.object({
  countyId: Joi.string().required().description('County ID'),
  parishId: Joi.string().required().description('Parish ID'),
  holdingsId: Joi.string().required().description('Holdings ID')
})

/**
 * @param {*} parameters
 * @returns {import('knex').Knex.Sql} the query builder for fetching units
 */
export function getUnitsQuery(parameters) {
  /**
   * validate parameters first
   */
  const { value, error } = GetUnitsSchema.validate(parameters)

  if (error) {
    throw new Error(`Invalid parameters: ${error.message}`)
  }

  const cph = `${value.countyId}/${value.parishId}/${value.holdingsId}`

  return query()
    .select(
      'cph',
      'location_id',
      'feature_name',
      'main_role_type',
      'person_family_name',
      'person_given_name',
      'organisation_name',
      'party_id',
      'asset_id',
      'asset_location_type',
      'asset_type',
      'animal_species_code',
      'animal_group_id_mch_ext_ref',
      'animal_group_id_mch_frm_dat',
      'animal_group_id_mch_to_dat',
      'animal_production_usage_code',
      'asset_involvement_type',
      'cph_type',
      'herdmark',
      'keeper_of_unit',
      'property_number',
      'postcode',
      'owner_of_unit'
    )
    .from('ahbrp.v_cph_customer_unit')
    .where('cph', `LIKE`, cph)
    .toSQL()
}
