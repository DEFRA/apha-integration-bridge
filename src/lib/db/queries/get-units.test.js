import { test, expect } from '@jest/globals'

import { getUnitsQuery } from './get-units.js'

test('returns the expected query for valid parameters', () => {
  const parameters = {
    countyId: '01',
    parishId: '02',
    holdingsId: '03'
  }

  const { sql, bindings } = getUnitsQuery(parameters)

  expect(sql).toMatchSnapshot()

  expect(bindings).toEqual(['01/02/03'])
})

test('throws if the parameters are invalid', () => {
  // @ts-expect-error - missing required parameters
  expect(() => getUnitsQuery({})).toThrow(/required/i)
})
