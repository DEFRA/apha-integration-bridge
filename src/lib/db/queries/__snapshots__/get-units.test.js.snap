// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the expected query for valid parameters 1`] = `"select cph, location_id, feature_name, main_role_type, person_family_name, person_given_name, organisation_name, party_id, asset_id, asset_location_type, asset_type, animal_species_code, animal_group_id_mch_ext_ref, animal_group_id_mch_frm_dat, animal_group_id_mch_to_dat, animal_production_usage_code, asset_involvement_type, cph_type, herdmark, keeper_of_unit, property_number, postcode, owner_of_unit from ahbrp.v_cph_customer_unit where cph like ?"`;
